### ISSUE TITLE
Helix test failure log makes it hard to diagnose failure

### ISSUE THREAD

**Issue by noahfalk** [dotnet/runtime#54957]
Working on [a recent PR](https://github.com/dotnet/runtime/pull/54333/checks?check_run_id=2951363176) the OSX tests failed. After following the link to [the log](https://helix.dot.net/api/2019-06-17/jobs/f690f8d0-f9b1-46b8-8dde-10a3ba276cbe/workitems/System.Diagnostics.DiagnosticSource.Tests/console) this is the information shown:

```
Console log: 'System.Diagnostics.DiagnosticSource.Tests' from job f690f8d0-f9b1-46b8-8dde-10a3ba276cbe workitem fad5a9d9-9dfc-4bc3-9708-bed82a1aa195 (osx.1014.amd64.open) executed on machine dci-mac-build-151
+ ./RunTests.sh --runtime-path /tmp/helix/working/B4910A08/p
----- start Wed Jun 30 03:27:42 PDT 2021 =============== To repro directly: =====================================================
pushd .
/tmp/helix/working/B4910A08/p/dotnet exec --runtimeconfig System.Diagnostics.DiagnosticSource.Tests.runtimeconfig.json --depsfile System.Diagnostics.DiagnosticSource.Tests.deps.json xunit.console.dll System.Diagnostics.DiagnosticSource.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing 
popd
===========================================================================================================
/private/tmp/helix/working/B4910A08/w/C4D20A3E/e /private/tmp/helix/working/B4910A08/w/C4D20A3E/e
  Discovering: System.Diagnostics.DiagnosticSource.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Diagnostics.DiagnosticSource.Tests (found 203 of 206 test cases)
  Starting:    System.Diagnostics.DiagnosticSource.Tests (parallel test collections = on, max threads = 12)
  Finished:    System.Diagnostics.DiagnosticSource.Tests
=== TEST EXECUTION SUMMARY ===
   System.Diagnostics.DiagnosticSource.Tests  Total: 257, Errors: 0, Failed: 0, Skipped: 0, Time: 13.556s
['System.Diagnostics.DiagnosticSource.Tests' END OF WORK ITEM LOG: Command timed out, and was killed]
```

The log ends saying that 'Command timed out and was killed'. If I had a wishlist of improvements to the scenario:
1. The timeout wouldn't have occured
2. The console output would make it obvious what code/process is in-progress and failed to complete. I think after some investigation xunit.console.dll hasn't yet exited, but I am uncertain and it took 20+ min of trying to learn how the automated scripts invoke each other to reach that guess.
3. When the timeout occured, dumps or other diagnostic data would be captured and linked to provide better odds of diagnosing the problem
4. This log would provide links or instructions how to find important assets such as:
  - The zip files containing files that helix deployed to run this work item. Most importantly [the one that contains RunTests.sh](https://helixde8s23ayyeko0k025g8.blob.core.windows.net/helix-job-acaf2ddc-07c4-459f-964e-0aec4db595bc10f1c118c904c3784/System.Diagnostics.DiagnosticSource.Tests.zip)
  - [Helix's log file for executing this work item](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-54333-merge-7af3c783f89349c58b/System.Diagnostics.DiagnosticSource.Tests/2b1612e3-d4eb-413f-a3a0-79d816388b61.log?sv=2019-07-07&se=2021-07-20T10%3A27%3A26Z&sr=c&sp=rl&sig=qimS3iudw%2B4KtBN5UMhsGm62f4TA9exKNrZ1VaXY3Tw%3D)

**Comment by noahfalk** [comment]
@hoyosjs - can you help me route this issue?

**Comment by hoyosjs** [comment]
> 1. The timeout wouldn't have occured

This one is not easily negotiable. We need to kill long running workitems. Weird that it says it ran for 13 seconds... However, if you see `System.Diagnostics.DiagnosticSource.Tests  Total: 257, Errors: 0, Failed: 0, Skipped: 0, Time: 13.556s`, that means tests ran all. We just had an unfortunate timeout after the fact, and there would be no dump. This is outside our tests. 

> 2. The console output would make it obvious what code/process is in-progress and failed to complete. I think after some investigation xunit.console.dll hasn't yet exited, but I am uncertain and it took 20+ min of trying to learn how the automated scripts invoke each other to reach that guess.

While this is definitely something that has room for improvement, we have two issues here. The first one is XUnit not showing what is happening. This is not as easy as it's highly parallel. Also, we can't easily modify the xunit console runner (and the verbose flag makes stuff really verbose and it's not useful). That being said, test timeouts and such are logged if we use `dotnet-test`

> 3. When the timeout occurred, dumps or other diagnostic data would be captured and linked to provide better odds of diagnosing the problem

We do collect dumps on crash, and CoreCLR does dump on timeout too if I am not wrong. This should light up for libraries once we start using `dotnet-test` (cc: @ViktorHofer, @safern)

> 4. This log would provide links or instructions how to find important assets such as ...

Console logs are from helix running. On the "Helix Test Logs (Beta)" tab you'd find the list of links to the cores. In library tests this even comes with a markdown on how to obtain the files. It basically shows this: https://github.com/dotnet/runtime/blob/main/eng/testing/debug-dump-template.md with with the job and workitem filled out. For yours it'd look like `runfo get-helix-payload --jobid aa7e2ad2-fa39-4e4d-8666-1fb4c70a6631 --workitems System.Diagnostics.DiagnosticSource.Tests -o .`. Running that would get everything to that was needed to run it - scripts, log, and all. 

Maybe the thing to consider here is to point to that tab if something in the test execution went wrong.

**Comment by hoyosjs** [comment]
Somehow that workitem lasted 45 min, out of which less than a min was test execution. As you said, XUnit is hung. Not sure how to make it obvious what's our part vs xunit. Maybe someone has an idea? In this case xunit and your test are in the same process. Unclear how we'd signal a non-exiting test harness. RunTests should also probably implement harness timeout on top of the test timeout for misbehaving infra.

**Comment by hoyosjs** [comment]
[Query](https://dataexplorer.azure.com/clusters/engsrvprod/databases/engineeringdata?query=H4sIAAAAAAAAAwvPL8r2LEnNLeaqUSjPSC1KVfDKT/JLzE1VsLVVUEpMNE81Skwx0k1LNLbUNUk1SdG1MDMz0zVMSzJJNjdINDMzNlRSSMxLUXArykzNS8mphGsNriwGGqvnkpmYnpdfXJKZXIzEDs4vLUpO1QtJLS4pVgIAS5R0hYMAAAA=)
```SQL
WorkItems
| where JobName == "aa7e2ad2-fa39-4e4d-8666-1fb4c70a6631" and FriendlyName == "System.Diagnostics.DiagnosticSource.Tests"
```

**Comment by noahfalk** [comment]
> > The timeout wouldn't have occured
> This one is not easily negotiable. We need to kill long running workitems.

I agree. Perhaps it wasn't well phrased. I mean "I wish that whatever the mystery thing is that made it run a long time is fixed, such that it would not have run a long time and not needed to timeout."

> > This log would provide links or instructions how to find important assets such as ...
> On the "Helix Test Logs (Beta)" tab you'd find

The crux of this wishlist item is that I want the info to be easily discoverable. If the links/instructions had been in this log I expect I would have found them right away. If the links/instructions are somewhere else then the value is lower based on how long it takes me to remember/discover what the other place is. Of course if I am the only one who has trouble finding them and for everyone else it is intuitive its fine to say that my experience is not representative and the solution is that I need to learn more about it. My default assumption is that if I had trouble then perhaps I am not the only one.

**Comment by noahfalk** [comment]
> Not sure how to make it obvious what's our part vs xunit. Maybe someone has an idea?

Two potential ideas:
1. I think you already said this one - RunTests.sh has its own timeout for xunit that is shorter than helix's overall timeout. This lets Runtests.sh print some explicit output "xunit process did exit after X minutes and was killed"
2. Structure the output in a way that strongly implies the nesting:

```
Running command: dotnet xunit.console.dll ...
All xunit command output shown inline between the { and } lines
{
  Discovering: System.Diagnostics.DiagnosticSource.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Diagnostics.DiagnosticSource.Tests (found 203 of 206 test cases)
  Starting:    System.Diagnostics.DiagnosticSource.Tests (parallel test collections = on, max threads = 12)
  Finished:    System.Diagnostics.DiagnosticSource.Tests
=== TEST EXECUTION SUMMARY ===
   System.Diagnostics.DiagnosticSource.Tests  Total: 257, Errors: 0, Failed: 0, Skipped: 0, Time: 13.556s

['System.Diagnostics.DiagnosticSource.Tests' END OF WORK ITEM LOG: Command timed out, and was killed]
